assert: add IsNotType

bartventer opened the pull request:
This commit introduces the `IsNotType` assertion, which is the inverse of the existing `IsType` assertion. It allows users to assert that an object is not of a specific type.

Additionally, minor documentation improvements were made.

## Summary
This PR adds a new assertion function, `IsNotType`, to the `testify/assert` package. It complements the existing `IsType` function by providing a way to assert that an object is not of a specific type.

## Changes
* Added the `IsNotType` function to the `assert` package.
* Wrote unit tests for `IsNotType` to ensure correctness.
* Updated documentation to include examples and usage for `IsNotType`.

## Motivation
The `testify/assert` package already provides an `IsType` function to check if an object is of a specific type. However, there was no built-in way to assert that an object is **not** of a specific type. This PR addresses that gap by introducing the `IsNotType` function, improving the library's completeness and usability.

## Example usage
```go
assert.IsNotType(t, &MyStruct{}, actualObject)
```

## Related issues
_N/A_

ccoVeille commented on assert/assertion_forward.go:
```diff
@@ -877,6 +899,8 @@ func (a *Assertions) IsType(expectedType interface{}, object interface{}, msgAnd
 
 // IsTypef asserts that the specified objects are of the same type.
+//
+//	assert.IsTypeff(t, &MyStruct{}, &MyStruct{})
```
```suggestion
//	assert.IsTypef(t, &MyStruct{}, &MyStruct{})
```

bartventer replied:
Addressed here https://github.com/stretchr/testify/pull/1730#issuecomment-2830300021

ccoVeille commented on assert/assertion_forward.go:
```diff
@@ -868,7 +868,29 @@ func (a *Assertions) IsNonIncreasingf(object interface{}, msg string, args ...in
+
 // IsType asserts that the specified objects are of the same type.
+//
+//	assert.IsTypef(t, &MyStruct{}, &MyStruct{})
```
```suggestion
//	assert.IsType(t, &MyStruct{}, &MyStruct{})
```

bartventer replied:
Addressed [here](https://github.com/stretchr/testify/pull/1730#issuecomment-2830300021)

ccoVeille commented on assert/assertions.go:
```diff
@@ -455,17 +455,34 @@ func NotImplements(t TestingT, interfaceObject interface{}, object interface{},
 // IsType asserts that the specified objects are of the same type.
-func IsType(t TestingT, expectedType interface{}, object interface{}, msgAndArgs ...interface{}) bool {
+//
+//	assert.IsTypef(t, &MyStruct{}, &MyStruct{})
```
```suggestion
//	assert.IsType(t, &MyStruct{}, &MyStruct{})
```

bartventer replied:
Addressed [here](https://github.com/stretchr/testify/pull/1730#issuecomment-2830300021)

ccoVeille reviewed (changes requested):
This sounds legitimate. I face the need this week too.

Please fix the small typos

bartventer commented:
I have addressed all the requested changes:
- Fixed the typo in the source file.
- Regenerated the rest of the files which reflect the corrections.

Let me know if there's anything else you'd like me to adjust.

ccoVeille commented on assert/assertions.go:
```diff
@@ -455,17 +455,34 @@ func NotImplements(t TestingT, interfaceObject interface{}, object interface{},
+	if h, ok := t.(tHelper); ok {
+		h.Helper()
+	}
+	return Fail(t, fmt.Sprintf("Object expected to not be of type %T, but was %T", theType, object), msgAndArgs...)
```
I feel like this wording could be improved

```suggestion
	return Fail(t, fmt.Sprintf("Object expected to not be of type %T", theType), msgAndArgs...)
```


```suggestion
	return Fail(t, fmt.Sprintf("Object type expected to be different than %T", theType), msgAndArgs...)
```

Both avoid the issue with current code in PR that would repeat

"not to be Whatever, but was Whatever"

bartventer replied:
I agree. Addressed in the [latest commit](https://github.com/stretchr/testify/pull/1730/commits/941eaaa16ae592dead1a5c3aaca727693b8f52c4).

ccoVeille reviewed (approved):
LGTM 👍

Thanks
